Add new Hausbus integration

hausbus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This is only the cover platform of the haus-bus.de integration. The rest will follow as soon this PR is accepted


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/40685
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [ x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x ] There is no commented out code in this PR.
- [ x] I have followed the [development checklist][dev-checklist]
- [x ] I have followed the [perfect PR recommendations][perfect-pr]
- [x ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

hausbus committed:
Add Hausbus integration

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
Merge branch 'add-hausbus' of https://github.com/hausbus/hacore.git into add-hausbus

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
lint

hausbus committed:
.

hausbus committed:
.

hausbus committed:
Fix after hassfest

hausbus committed:
Fix after hassfest

hausbus committed:
code checks

hausbus committed:
code checker

hausbus committed:
code checks

hausbus committed:
Merge branch 'dev' into add-hausbus

hausbus committed:
.

hausbus committed:
.

hausbus committed:
prepared cover platform

hausbus commented on .pre-commit-config.yaml:
```diff
@@ -12,6 +12,7 @@ repos:
     hooks:
       - id: codespell
         args:
+          - --ignore-words=.codespellignore
```
reverted

hausbus commented on .settings/org.eclipse.core.resources.prefs:
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
removed

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
fixed pr tasks

hausbus commented on node_modules/.bin/prettier:
```diff
@@ -0,0 +1 @@
+../prettier/bin/prettier.cjs
```
removed

hausbus commented on package.json:
```diff
@@ -0,0 +1,5 @@
+{
```
removed

hausbus committed:
Entferne Datei aus PR

joostlek commented on .pre-commit-config.yaml:
This shouldn't be deleted

joostlek replied:
This shouldn't be updated either

joostlek reviewed (changes requested):
We should also remove the services and device actions from this PR.

hausbus commented:
> We should also remove the services and device actions from this PR.

why? the services and actions are needed for the cover entity

hausbus commented:
> The cover has methods to work on its own right?

sure but cover does not support toggle of direction or am I wrong ?

joostlek commented:
Cover has a built in feature to toggle the cover based on the information that the cover exposes

hausbus committed:
removed toggle service and action from cover

hausbus committed:
reverted to original state

hausbus committed:
Merge branch 'dev' into hausbus-cover

hausbus committed:
Merge branch 'dev' into hausbus-cover

hausbus committed:
removed device_actions.py

hausbus commented:
> We should also remove the services and device actions from this PR.
@joostlek I removed device_actions.py now but if I go to Automations and open the possible actions types for a cover TOGGLE action is not offered. Do I miss something? I understood that the cover entity automatically supports toggle using the last reported movement direction but how can it be supported as an automation actions.
Second question. Do you know why this PR is still labled with cla-error although I fixed the problem and cla-bot seooms to be happy ? Thank you!

hausbus committed:
added CoverEntityFeature.TOGGLE

hausbus committed:
Merge branch 'hausbus-cover' of https://github.com/hausbus/hacore.git into hausbus-cover

hausbus committed:
readded cover.toggle device action due to missing automation ui action
entry

hausbus committed:
.

hausbus committed:
readded cover.toggle device action again to make it appear in automation ui because calling default entity function cover.toggle

hausbus committed:
Merge branch 'dev' into hausbus-cover

joostlek commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,124 @@
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class HausbusConfig:
+    """Class for Hausbus ConfigEntry."""
+
+    gateway: HausbusGateway
+
+
+HausbusConfigEntry = ConfigEntry[HausbusConfig]
```
If we're going to only store the gateway in the config entry, might as well skip the dataclass

joostlek replied:
```suggestion

type HausbusConfigEntry = ConfigEntry[HausbusConfig]
```

joostlek commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,124 @@
+async def async_setup_entry(hass: HomeAssistant, entry: HausbusConfigEntry) -> bool:
+    """Set up Haus-Bus integration from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN]["entity_info"] = {}
```
I am not sure why we are still storing things in the hass.data, we can store everything in entry.runtime_data

joostlek commented on homeassistant/components/hausbus/services.yaml:
This shouldn't be added right now

hausbus replied:
why not ? I need the set_configuration service

joostlek commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN]["entity_info"] = {}
+
+    try:
+        gateway = HausbusGateway(hass, entry)
+    except Exception as err:
+        raise ConfigEntryNotReady(f"Setup failed: {err}") from err
+
```
Creating this object doesn't raise right?

hausbus replied:
I only added this because of:

Integration hausbus:
* [ERROR] [QUALITY_SCALE] [test-before-setup] Integration does not raise one of {'ConfigEntryNotReady', 'ConfigEntryAuthFailed', 'ConfigEntryError'}
* [ERROR] [QUALITY_SCALE] Please check the documentation at https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/test-before-setup/

joostlek commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Creates a button to manually start device discovery
+    hass.async_create_task(gateway.createDiscoveryButton())
```
It would be nice if we can use snake_case for all methods in the class in the library

joostlek commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,124 @@
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def async_remove_config_entry_device(
+    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: DeviceEntry
+) -> bool:
+    """Handle removal of a device from the integration."""
+
+    gateway = config_entry.runtime_data.gateway
+
+    identifiers = device_entry.identifiers
+    for domain, device_id in identifiers:
+        if domain == DOMAIN:
+            return bool(await gateway.removeDevice(device_id))
+
+    return False
```
why do we need this?

hausbus replied:
To have the DELETE_DEVICE link in the front end

hausbus replied:
<img width="335" height="248" alt="image" src="https://github.com/user-attachments/assets/d074145d-b007-4661-82e0-c0ddb257911e" />

hausbus committed:
Revert .pre-commit-config.yaml to dev version

hausbus committed:
PR changes

hausbus committed:
fixed bug caused by entity separation for PR

hausbus commented:
@joostlek  Sorry to bother you. Is it on me to do something ?

hausbus committed:
Update .pre-commit-config.yaml

hausbus committed:
Update .pre-commit-config.yaml

hausbus commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,110 @@
+LOGGER = logging.getLogger(__name__)
+
+
+HausbusConfigEntry = ConfigEntry[HausbusGateway]
```
added

hausbus commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,110 @@
+
+    async def discover_devices(call: ServiceCall):
+        entries = hass.config_entries.async_entries(DOMAIN)
+        if not entries:
+            raise HomeAssistantError("No Hausbus-Gateway available")
+
+        LOGGER.debug("Search devices service called")
+        gateway = entries[0].runtime_data
+        gateway.home_server.searchDevices()
+
+    hass.services.async_register(DOMAIN, "discover_devices", discover_devices)
+
+    async def reset_service(call):
+        entries = hass.config_entries.async_entries(DOMAIN)
+        if not entries:
+            raise HomeAssistantError("No Hausbus-Gateway available")
+
+        LOGGER.debug("call  %s ", call)
+
+        device_id = call.data.get("device_id")
+        if not device_id and getattr(call, "target", None):
+            device_id = getattr(call.target, "device_ids", [])
+
+        if not device_id:
+            raise HomeAssistantError("device_id missing")
+
+        device_id = device_id[0]
+
+        LOGGER.debug("Reset device %s called", device_id)
+        gateway = entries[0].runtime_data
+        try:
+            gateway.resetDevice(device_id)
+        except Exception as err:
+            raise HomeAssistantError(
+                f"Failed to reset device {device_id}: {err}"
+            ) from err
+
+    hass.services.async_register(DOMAIN, "reset_device", reset_service)
+
+    return True
```
@joostlek Hi, can you explain why? I mean, it does not help if we integrate only one entity but you do not have the very basic functions like searching new devices or reset a device. I understand that we should add one only platform at a time but why do we have to skip all needed services ?

joostlek replied:
1. But if we do it automatically we don't have to manually trigger it either. People can reload the integration if they want to have it now now
2. A button is also scriptable

hausbus replied:
But I dont want the User to provide a reset Button. What is the Problem with the Service itself? And why can't I provide a Button to manually start a device search? Restarting the Integration is absolutely unconfortable

hausbus replied:
I now removed reset service and delete device service

hausbus commented on homeassistant/components/hausbus/__init__.py:
```diff
@@ -0,0 +1,110 @@
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+async def async_remove_config_entry_device(
+    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: DeviceEntry
+) -> bool:
+    """Handle removal of a device from the integration."""
+
+    gateway = config_entry.runtime_data
+
+    identifiers = device_entry.identifiers
+    for domain, device_id in identifiers:
+        if domain == DOMAIN:
+            return bool(await gateway.removeDevice(device_id))
+
+    return False
```
@joostlek same as above. I think it is really needed to have the possibility to remove a device in the HA frontend

joostlek replied:
Yes, that's fine, but I think that this isn't a primary feature for this PR and we can add this in a separate PR and have separate discussion on this in there

hausbus replied:
I can remove it but I really dont understand this approach. Don't we need a bare minimum to start? Only one Entity, OK, but search for Services and deleted Services is really needed. Isn`t it? And this 5 lines of Code do Not add so much complexity that the PR is Not handable anymore

hausbus commented on homeassistant/components/hausbus/device.py:
```diff
@@ -0,0 +1,99 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class HausbusDevice:
+    """Common base class for Haus-Bus devices."""
+
+    def __init__(
+        self, device_id: str, sw_version: str, hw_version: str, firmware_id: EFirmwareId
+    ) -> None:
+        """Set up Haus-Bus device."""
+        self.device_id = device_id
+        self.manufacturer = "Haus-Bus.de"
+        self.model_id = "Controller"
```
@joostlek It is just a helper class to organize the haus-bus devices. What is the problem having it ?

joostlek replied:
No, I think that the object itself is not needed as the task is too small and mergable with the other objects that exist like the entity

hausbus replied:
But we have one device which has X entities. Why should I duplicate the device information into each Entity?

hausbus commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,351 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class HausbusGateway(IBusDataListener):
```
@joostlek Well, this is the main class that takes care to dynamically create entites for all queried devices and their functions. In generall the class is not too complex. What would you suggest to simplify ?

hausbus replied:
Moved device specific internals to library

joostlek replied:
> Do you expect that somebody else then me will ever have to maintain this Integration?

Yes, we should always expect that. If you suddenly have different priorities and one of your users has a bug and wants to fix that, they should be able to understand the code.

hausbus committed:
pr changes

hausbus committed:
Merge branch 'hausbus-cover' of https://github.com/hausbus/hacore.git
into hausbus-cover

hausbus committed:
.

hausbus commented on homeassistant/components/hausbus/icons.json:
@joostlek How to handle this checker errors without icons.json ?
[ERROR] [SERVICES] Service discover_devices has no icon in icons.json.
[ERROR] [SERVICES] Service reset_device has no icon in icons.json.
[ERROR] [SERVICES] Service cover_set_configuration has no icon in icons.json.

hausbus commented on homeassistant/components/hausbus/cover.py:
```diff
@@ -0,0 +1,200 @@
+            self._attr_extra_state_attributes["open_time"] = data.getOpenTime()
+            self._attr_extra_state_attributes["invert_direction"] = (
+                data.getOptions().isInvertDirection()
+            )
```
What is the best way to show this configuration values to the user without that they appear in automations ?

hausbus replied:
Ok, I will do

hausbus replied:
removed attributes

hausbus commented:
@joostlek 
Side question: what do you estimate how long this integration will take to be finished? I have 6 entities left and a lot of side stuff. I thought that if I fulfill all guidelines and all pre-commit checks then it will be OK. Why is there so much additional focus in Style? I parallel a lot of users are trsting the complete integration via HACS.

hausbus commented on homeassistant/components/hausbus/cover.py:
```diff
@@ -0,0 +1,200 @@
+            ),
+        },
+        "async_cover_set_configuration",
+    )
```
removed service

hausbus committed:
Changes requested by HA reviewer

joostlek commented on homeassistant/components/hausbus/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+STEP_USER_SCHEMA = vol.Schema({})
+
+
+class ConfigFlow(IBusDataListener, config_entries.ConfigFlow, domain=DOMAIN):
```
1. you did not push
2. We should still connect to the device. But my point is that we should not inherit both this interface and a configFlow in the same object. Instead create a separate object that does what the config flow should do

hausbus committed:
PR preparation

hausbus commented:
@joostlek I now moved much device specific code to the lib and removed everything that is not needed for the first PR. Thanks for checking again

hausbus committed:
Merge branch 'dev' into hausbus-cover

hausbus committed:
overworked device discovery

hausbus committed:
Merge branch 'hausbus-cover' of https://github.com/hausbus/hacore.git into hausbus-cover

hausbus committed:
.

hausbus committed:
.

hausbus committed:
.

hausbus committed:
overworked device discovery

joostlek reviewed (changes requested):
I think the integration is heading in a right direction, I left some more architectural things to think about

hausbus committed:
Apply suggestion from @joostlek

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

hausbus committed:
Apply suggestion from @joostlek

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

hausbus commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,181 @@
+        # callbacks to dynamically register new entities discovered by this gateway
+        self._new_channel_listeners: dict[
+            str, Callable[[HausbusEntity], Coroutine[Any, Any, None]]
+        ] = {}
```
refactored and moved as class dict to HausBusEntity

hausbus commented on homeassistant/components/hausbus/manifest.json:
```diff
@@ -0,0 +1,22 @@
+    "cover",
+    "button",
+    "number"
+  ],
```
removed

hausbus commented on tests/components/hausbus/test_config_flow.py:
ok, lets check later

hausbus committed:
PR changes

hausbus committed:
Revert changes to requirements_test_pre_commit.txt

hausbus committed:
Revert changes to requirements_test_pre_commit.txt

hausbus committed:
removed device_action

hausbus commented on homeassistant/components/hausbus/device_action.py:
Driving a Cover in the opposite direction. This is not selectable as automation Action although the Cover Entity itself provides this as a logical function. For that reason I added it as an Action because then it was selectable

joostlek reviewed (changes requested):
The PR is now a good size to review, it has become a lot easier to understand what is going on

hausbus committed:
removed redundant logging

hausbus committed:
removed redundant logging

hausbus committed:
removed unnessessary constant

hausbus commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,140 @@
+        """Registers entity with the prior remembered callback."""
+        cb = cls._domain_add_entity_callbacks.get(entity.get_domain())
+        if cb is not None:
+            cb([entity])
```
I have to remember the AddConfigEntryEntitiesCallback somewhere because I cannot register the entities during async_setup_entry since they have to be created during runtime dynamically.
I now moved the dict to gateway.py to fulfil your requirement that entities should not care for the registration

hausbus commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,140 @@
+
+        # add type information for individual actions
+        self.hass.data.setdefault(DOMAIN, {}).setdefault("device_types", {})
+        self.hass.data[DOMAIN]["device_types"][self.entity_id] = self.__class__.__name__
```
we probably come back to that later. I removed it for now

hausbus committed:
pr

hausbus committed:
pr

hausbus committed:
.

hausbus committed:
removed config receiver

hausbus commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,140 @@
+    async def _wait_for_configuration(self):
+        """Waits until configuration is received."""
+        while not self._configuration:
+            await asyncio.sleep(0.1)
```
I removed it

hausbus committed:
pr

hausbus committed:
pr

hausbus committed:
pr

hausbus commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,160 @@
+            self.hass,
+            f"hausbus_update_{object_id.getValue()}",
+            data,
+        )
```
done

joostlek commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,170 @@
+class HausbusGateway(IBusDataListener):
+    """Manages a Haus-Bus gateway."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```
use the extended config entry here. You might need to move the `type` to this file

joostlek reviewed (changes requested):
Going for a smaller review here as I want to address some small things, if these are resolved we can improve from there

hausbus commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,170 @@
+                if new_entity is not None:
+                    asyncio.run_coroutine_threadsafe(
+                        self.register_entity(new_entity), self.hass.loop
+                    ).result()
```
I now tried to inform about dynamically detected new channels with a dispatcher that the platform registers to and eventually creates corresponding entities. I hope this is now what you want to see

hausbus committed:
pr

hausbus commented on homeassistant/components/hausbus/gateway.py:
```diff
@@ -0,0 +1,170 @@
+from .cover import HausbusCover, Rollladen
+from .entity import HausbusEntity
+
+LOGGER = logging.getLogger(__name__)
```
fixed

hausbus commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,140 @@
+        self._attr_name = "to be overridden"
+        self._attr_unique_id = "to be overridden"
+
+        if channel is not None:
```
why would WHAT not be there?

hausbus replied:
Not sure, if I understood your question. Is there actually still something wrong or open ?

hausbus committed:
added all rules and grades

hausbus commented:
I also have two questions pending:

1. **How can a user toggle the direction of a cover within automations**?
    If I add an action in a automation and choose Device then I can only choose for cover: Open, Close, Stop, setPosition but no    
    toggle. For that reason I added toggle as a service. How can I provide toggle without a service?

2. **How to show configuration data and provide configurations**
    For a cover I want to show the configured open_time, close_time and if the cover is inverted or not.
    What is the best way to show this information to a user and let him configure this parameter ?

Thanks

joostlek commented on homeassistant/components/hausbus/cover.py:
```diff
@@ -0,0 +1,167 @@
+            | CoverEntityFeature.CLOSE
+            | CoverEntityFeature.STOP
+            | CoverEntityFeature.SET_POSITION
+        )
```
Can be set outside of the constructor

joostlek commented on homeassistant/components/hausbus/entity.py:
```diff
@@ -0,0 +1,84 @@
+        self._unsub_dispatcher: Any = None
+        self._domain = domain
+
+        if channel is not None:
```
Again, channel can't be None according to the typing

joostlek commented on homeassistant/components/hausbus/quality_scale.yaml:
```diff
@@ -0,0 +1,65 @@
+    status: done
+    comment: |
+      Registers the `hausbus.discover_devices` service, which calls `searchDevices` on the gateway.
+  appropriate-polling: done
```
It's not polling right?

hausbus committed:
Update homeassistant/components/hausbus/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

hausbus commented on homeassistant/components/hausbus/cover.py:
```diff
@@ -0,0 +1,167 @@
+    @property
+    def is_closing(self) -> bool | None:
+        """Returns true if cover is actually closing."""
+        return self._is_closing
```
fixed
